firedrake/adjoint/blocks.py

```diff
@@ -247,3 +249,38 @@ def recompute_component(self, inputs, block_variable, idx, prepared):
         mesh = vector.function_space().mesh()
         mesh.coordinates.assign(vector, annotate=False)
         return mesh._ad_create_checkpoint()
+
+
+class PointwiseOperatorBlock(Block, Backend):
+    def __init__(self, point_op, *args, **kwargs):
+        super(PointwiseOperatorBlock, self).__init__()
+        self.point_op = point_op
+        for c in self.point_op.ufl_operands:
+            coeff_c = ufl.algorithms.extract_coefficients(c)
+            for ci in coeff_c:
+                self.add_dependency(ci, no_duplicates=True)
+        self.add_dependency(self.point_op, no_duplicates=True)
+
+    def prepare_evaluate_adj(self, inputs, adj_inputs, relevant_dependencies):
+        N, ops = inputs[-1], inputs[:-1]
+        return N._ufl_expr_reconstruct_(*ops)
+
+    def evaluate_adj_component(self, inputs, adj_inputs, block_variable, idx, prepared=None):
+        if self.point_op == block_variable.output:
+            # We are not able to calculate derivatives wrt initial guess.
+            return None
+
+        q_rep = block_variable.saved_output
+        N = prepared
```

Comment:
Can you not just use `prepared`?

Reply:
I have replaced `prepared` by `N` as it seems more intuitive to see what is happening